src/core/server/elasticsearch/client/client_wrapper.ts

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Client } from '@elastic/elasticsearch';
+import { TransportRequestOptions } from '@elastic/elasticsearch/lib/Transport';
+import { Headers } from '../../http/router';
+import { ClientFacade } from './client_facade';
+
+export const getClientFacade = (client: Client, headers: Headers = {}): ClientFacade => {
+  const addHeaders = (options?: TransportRequestOptions): TransportRequestOptions => {
+    if (!options) {
+      return {
+        headers,
+      };
+    }
+    // TODO: do we need to throw in case of duplicates as it was done
+    //       in legacy? - src/core/server/elasticsearch/scoped_cluster_client.ts:L88
+    return {
+      ...options,
+      headers: {
+        ...options.headers,
+        ...headers,
+      },
+    };
+  };
+
+  return {
+    bulk: (params, options) => client.bulk(params, addHeaders(options)),
+    asyncSearch: {
+      delete: (params, options) => client.asyncSearch.delete(params, addHeaders(options)),
+      get: (params, options) => client.asyncSearch.get(params, addHeaders(options)),
+      submit: (params, options) => client.asyncSearch.submit(params, addHeaders(options)),
+    },
+  };
```

Comment:
I started by creating a `class ClientWrapper implements ClientFacade`. However
- You can't define nested properties in a class definition, meaning that I would have to create subtypes / subclasses for all nested call structures such as `asyncSearch` here.
- Implementing the methods on a class forces to have an explicit signature and parameter types, which is not the case when just returning an inline object. This is a **huge** difference in term of size and readability.

Reply:
that's the case when it's easier to infer type manually. as you did here or via `const wrapper: ClientFacade = { ...`